src/rpc.rs

```diff
@@ -1,105 +1,184 @@
+    // token) and Room, which we should replace by encoding the room token in the endpoint.
     pub headers: HashMap<String, String>,
+    // For new, token-less requests; requests without these should be considered deprecated.
+    pub ed25519_pubkey: Option<String>, // Ed25519 pubkey, in hex (will be used to derive the Session key by converting and prepending 05)
+    pub nonce: Option<String>, // Arbitrary string; must be different on each request
+    pub signature: Option<String>, // Ed25519 signature (in base64 or hex) of (method || endpoint || body || nonce)
 }
 
 pub const MODE: Mode = Mode::OpenGroupServer;
 
-pub async fn handle_rpc_call(rpc_call: RpcCall) -> Result<Response, Rejection> {
+// Gets a user from a reflected auth token.  Returns None if there is no auth token, User if there
+// is a parseable auth token, and an error for anything else.
+fn get_user_from_auth_header(conn: &rusqlite::Connection, rpc: &RpcCall) -> Result<Option<User>, Error> {
+    if let Some(auth_token_str) = rpc.headers.get("Authorization") {
+        return Ok(Some(handlers::get_user_from_token(conn, auth_token_str)?));
+    }
+    return Ok(None);
+}
+
+// FIXME TODO - these calls using warp::Rejection as an error type are Doing It Wrong:
+// warp::Rejection means "decline this handler, try another one" but everywhere we are using it we
+// mean "return an error".
+//
+// c.f. https://github.com/seanmonstar/warp/issues/388
+
+pub async fn handle_rpc_call(mut rpc_call: RpcCall) -> Result<Response, Rejection> {
+
+    let have_sig = rpc_call.ed25519_pubkey.is_some();
+    if rpc_call.nonce.is_some() != have_sig || rpc_call.signature.is_some() != have_sig  {
+        warn!("Invalid request: all or none of {{ed25519_pubkey, nonce, signature}} must be provided");
+        return Err(Error::InvalidRpcCall.into());
+    }
+
+    let mut user: Option<User> = None;
+    if have_sig {
+        let (edpk, _xpk, sessionid) = crypto::get_pubkeys(rpc_call.ed25519_pubkey.as_ref().unwrap())?;
+
+        let nonce = rpc_call.nonce.as_ref().unwrap();
+        // TODO FIXME: reject recent pubkey/nonce combinations.
+
+        let mut sig_bytes: [u8; 64] = [0; 64];
+        sig_bytes.copy_from_slice(&handlers::decode_hex_or_b64(rpc_call.signature.as_ref().unwrap(), 64)?[0..64]);
+        let sig = ed25519_dalek::Signature::new(sig_bytes);
+
+        if let Err(sigerr) = handlers::verify_signature(&edpk, &sig,
+            &vec![rpc_call.endpoint.as_bytes(), rpc_call.method.as_bytes(), rpc_call.body.as_bytes(), nonce.as_bytes()]) {
+            warn!("Signature verification failed for request from {}", sessionid);
+            return Err(sigerr.into());
+        }
+
+        user = Some(handlers::insert_or_update_user(&*storage::get_conn()?, &sessionid)?);
+
+        // Check for a global ban, and if so, terminate the request right away.
+        if user.as_ref().unwrap().banned {
```

Comment:
Could you not check if `user` is banned before wrapping it in `Some`? I can see that it is safe to unwrap in this case, but it does add a bit of extra friction when reading the code.